packages/backend/src/peripherals/database/ForcedTransactionsRepository.ts

```diff
@@ -0,0 +1,419 @@
+): ForcedTransaction {
+  if (event.transactionType === 'withdrawal' && event.eventType === 'mined') {
+    return {
+      type: 'withdrawal',
+      status: 'mined',
+      hash: event.transactionHash,
+      lastUpdate: event.timestamp,
+      ...pick(event, 'amount', 'positionId', 'publicKey'),
+    }
+  }
+  if (
+    transaction.type === 'withdrawal' &&
+    event.transactionType === 'withdrawal' &&
+    event.eventType === 'verified'
+  ) {
+    return {
+      ...transaction,
+      status: 'verified',
+      stateUpdateId: event.stateUpdateId,
+      lastUpdate: event.timestamp,
+    }
+  }
+  if (event.transactionType === 'trade' && event.eventType === 'mined') {
+    return {
+      type: 'trade',
+      status: 'mined',
+      hash: event.transactionHash,
+      ...pick(
+        event,
+        'collateralAmount',
+        'publicKeyA',
+        'publicKeyB',
+        'positionIdA',
+        'positionIdB',
+        'syntheticAssetId',
+        'isABuyingSynthetic',
+        'collateralAmount',
+        'syntheticAmount'
+      ),
+      lastUpdate: event.timestamp,
+    }
+  }
+  if (
+    transaction.type === 'trade' &&
+    event.transactionType === 'trade' &&
+    event.eventType === 'verified'
+  ) {
+    return {
+      ...transaction,
+      status: 'verified',
+      stateUpdateId: event.stateUpdateId,
+    }
+  }
+  throw new Error(
+    'Unknown type combination: ' +
+      transaction.toString() +
+      ' ' +
+      event.toString()
+  )
+}
+
+function reduceEventsToTransaction(events: EventRecord[]): ForcedTransaction {
+  const sorted = events.sort(
+    (a, b) => Number(a.timestamp) - Number(b.timestamp)
+  )
+  return sorted.reduce(applyEvent, {} as ForcedTransaction)
+}
+
+function eventRowsToTransactions(rows: EventRow[]): ForcedTransaction[] {
+  const events = rows.map(toRecord)
+  const eventsGroupedByTransactionHash = Object.values(
+    groupBy(events, (e) => e.transactionHash)
+  ).sort((a, b) => Number(b[0].timestamp) - Number(a[0].timestamp))
+
+  return eventsGroupedByTransactionHash.map(reduceEventsToTransaction)
+}
+
+export class ForcedTransactionsRepository {
+  constructor(private knex: Knex, private logger: Logger) {
+    this.logger = logger.for(this)
+  }
+
+  async addEvents(candidates: EventRecordCandidate[]): Promise<void> {
+    if (candidates.length === 0) {
+      return
+    }
+    const rowCandidates = candidates.map(recordCandidateToRow)
+    await this.knex('forced_transaction_events').insert(rowCandidates)
+    this.logger.debug({
+      method: 'addEvents',
+      count: candidates.length,
+    })
+  }
+
+  async getTransactionHashesByMinedEventsData(
+    datas: (TradeMinedEventData | WithdrawalMinedEventData)[]
+  ): Promise<(Hash256 | undefined)[]> {
+    if (datas.length === 0) {
+      return []
+    }
+    const hashes = datas.map(toJsonString).map(hashData)
+    const events = await this.knex('forced_transaction_events').whereRaw(
+      "data->>'hash' = any(?)",
```

Comment:
@sin6pi7 @sz-piotr what if there is more than one transaction with the same data hash?

Reply:
@sz-piotr It is actually already taken care of - only one is returned (look at the `events.find` call in `hashes.map` function). However, the order of events is not defined - given two mined events with the same data shall we care about which one we choose? If so, what would be the logic?